cpp/src/dictionary/detail/concatenate.cu

```diff
@@ -0,0 +1,268 @@
+             cudaStream_t stream,
+             rmm::mr::device_memory_resource* mr)
+  {
+    auto keys_view     = column_device_view::create(all_keys, stream);
+    auto d_all_keys    = *keys_view;
+    auto indices_view  = column_device_view::create(all_indices, stream);
+    auto d_all_indices = *indices_view;
+
+    auto all_itr = thrust::make_transform_iterator(
+      thrust::make_counting_iterator<size_type>(0),
+      [d_all_keys, d_offsets, d_map_to_keys, d_all_indices] __device__(size_type idx) {
+        if (d_all_indices.is_null(idx)) return Element{};
+        size_type index =
+          d_all_indices.template element<int32_t>(idx) + d_offsets[d_map_to_keys[idx]].first;
+        return d_all_keys.template element<Element>(index);
+      });
+
+    auto new_keys_view = column_device_view::create(new_keys, stream);
+    auto d_new_keys    = *new_keys_view;
+    auto keys_itr      = thrust::make_transform_iterator(
+      thrust::make_counting_iterator<size_type>(0),
+      [d_new_keys] __device__(size_type idx) { return d_new_keys.template element<Element>(idx); });
+
+    // compute new indices values into output column
+    auto result = make_numeric_column(
+      data_type{type_id::INT32}, all_indices.size(), mask_state::UNALLOCATED, stream, mr);
+    auto d_result = result->mutable_view().data<int32_t>();
+    thrust::lower_bound(rmm::exec_policy(stream)->on(stream),
+                        keys_itr,
+                        keys_itr + new_keys.size(),
+                        all_itr,
+                        all_itr + all_indices.size(),
+                        d_result,
+                        thrust::less<Element>());
+    return result;
+  }
+
+  template <typename Element>
+  typename std::enable_if_t<!cudf::is_relationally_comparable<Element, Element>(),
+                            std::unique_ptr<column>>
+  operator()(column_view const&,
+             column_view const&,
+             column_view const&,
+             offsets_pair const*,
+             size_type const*,
+             cudaStream_t stream,
+             rmm::mr::device_memory_resource*)
+  {
+    CUDF_FAIL("list_view as keys for dictionary not supported");
+  }
+};
+
+}  // namespace
+
+std::unique_ptr<column> concatenate(std::vector<column_view> const& columns,
+                                    cudaStream_t stream,
+                                    rmm::mr::device_memory_resource* mr)
+{
+  if (columns.size() == 0) return make_empty_column(data_type{type_id::DICTIONARY32});
+
+  // concatenate the keys (and check the keys match)
+  compute_children_offsets_fn child_offsets_fn{columns};
+  auto keys_type = child_offsets_fn.get_keys_type();
+  std::vector<column_view> keys_views(columns.size());
+  std::transform(columns.begin(), columns.end(), keys_views.begin(), [keys_type](auto cv) {
+    auto dict_view = dictionary_column_view(cv);
+    // empty column may not have keys so we create an empty column_view place-holder
```

Comment:
Is this still true with the latest decision we made about ensuring compound columns always have children even when empty? I'd think the same problems would apply to a childless dictionary column as a childless list column, you wouldn't know what it is a dictionary _of_.

Reply:
It is still true right now. Regardless if that change occurs or not, this code will work either way. This simple allows the code to continue if an empty dictionary column is encountered.
Reference #5816